File: backend/app/schemas/analytics.py
"""
analytics.py

Pydantic schemas for exercise progress analytics and chart filters.
"""

from __future__ import annotations

from datetime import date as date_type
from typing import Literal

from pydantic import BaseModel, Field


ProgressMetric = Literal["weight", "reps", "volume"]


class ProgressPoint(BaseModel):
    """One recorded point for an exercise set on a given date."""

    date: date_type
    weight: float = Field(..., ge=0)
    reps: int = Field(..., ge=0)
    volume: float = Field(..., ge=0)


class ProgressSeries(BaseModel):
    """Progress points for a single set number across workout dates."""

    set_number: int = Field(..., gt=0)
    points: list[ProgressPoint] = Field(default_factory=list)


class ExerciseProgressOut(BaseModel):
    """Analytics response for one exercise with one or more set-number series."""

    exercise_id: int
    exercise_name: str
    series: list[ProgressSeries] = Field(default_factory=list)

File: frontend/src/api/analytics.ts
import { apiClient } from "./client";

const ANALYTICS_PREFIX = "/api/v1/analytics";

export type ProgressMetric = "weight" | "reps" | "volume";

export interface ProgressPoint {
  date: string;
  weight: number;
  reps: number;
  volume: number;
}

export interface ProgressSeries {
  set_number: number;
  points: ProgressPoint[];
}

export interface ExerciseProgressOut {
  exercise_id: number;
  exercise_name: string;
  series: ProgressSeries[];
}

export interface ProgressQueryParams {
  start_date?: string;
  end_date?: string;
  set_number?: number;
}

export interface ProgressChartParams extends ProgressQueryParams {
  metric: ProgressMetric;
}

export async function getProgress(
  exerciseId: number,
  params?: ProgressQueryParams
): Promise<ExerciseProgressOut> {
  const { data } = await apiClient.get<ExerciseProgressOut>(
    `${ANALYTICS_PREFIX}/progress/${exerciseId}`,
    { params: params ?? {} }
  );
  return data;
}

export async function getProgressChartObjectUrl(
  exerciseId: number,
  params: ProgressChartParams
): Promise<string> {
  const response = await apiClient.get<Blob>(`${ANALYTICS_PREFIX}/progress/${exerciseId}/chart`, {
    params,
    responseType: "blob",
  });
  return URL.createObjectURL(response.data);
}
